templates/ruleset.epp

```diff
@@ -11,43 +11,84 @@ ruleset (name="<%= $ruleset_name %>"
 <%}-%>
 ) {
 <% $rules.each |$rule| { -%>
-<% $rule.each |$key, $params| { -%>
-<% if $key == 'set' { -%>
-<%- $params.each |$var, $value| { -%>
+<%   $rule.each |$key, $params| { -%>
+<%     if $key == 'filter' { -%>
+<%       $params.each |$conditional, $block| { -%>
+<%         if $conditional == 'else' { -%>
+  else {
+<%         } else { -%>
+  <%= "${conditional} (${block['filter']} ${block['operator']} '${block['expression']}') " -%> then {
+<%         }-%>
+<%         $block['tasks'].each |$task, $value| { -%>
+<%           if $task == 'call' { -%>
+    call <%= $value %>
+<%           } elsif $task == 'stop' { -%>
+    stop
+<%           } elsif $task == 'set' { -%>
+<%             $task.each |$var, $value| { -%>
+    set $.<%= $var %> = <%= $value %><%-%>;
+<%             } -%>
+<%           } elsif $task == 'action' { -%>
+<%       if $params['facility'] { -%>
+<%= sprintf( '%-30s %s%s%s', $params['facility'], 'action(type="',$params['type'],'" ' )%>
```

Comment:
Refactor at some point to have this template call the `action` template for action generation rather than copying code multiple times?

Reply:
What guarantees the correct ordering of tasks if they are in a hash? Also what:
```
} elsif $task == 'set' { $task.each ....
```
How is that supposed to work?

What do you do if you want to call multiple functions?

I think it would help to add some custom types to help enforce the structure, since this template is quite complicated.

Reply:
The template reads in from a hash-based parameter and checks keys within the hash. So:

```yaml
rsyslog::server::rulesets:
  ruleset_msg_check_for_error:
    rules:
      - expression_filter:
          property: 'msg'
          operator: 'contains'
          value: 'error'
          tasks:
            call: 'ruleset.action.error'
            stop: true
```

Everything in the above example is loaded in as it is read in. You can define more calls, actions, etc. The only thing that can't be defined twice is the stop. 

Rsyslog rulesets and expressions are highly dependent  on the order the content within it is defined and there's no real way to pre-define that ordering because a `call`, `action`, `set`, `lookup`, etc can be in any order based on what the person writing the config needs.

That said, the code needs to be refactored at some point to provide cleaner templates and, potentially, more clearly defined parameters for each feature. That said, Rsyslog is an incredibly complex application and we need to make sure this module continues to support the standard basic use as well as the more complex uses of it.